--- a/kotlin/internal/jvm/jvm.bzl
+++ b/kotlin/internal/jvm/jvm.bzl
@@ -173,6 +173,12 @@
         "srcs": attr.label_list(
             doc = """The list of source files that are processed to create the target, this can contain both Java and Kotlin
         files. Java analysis occurs first so Kotlin classes may depend on Java classes in the same compilation unit.""",
             default = [],
             allow_files = [".srcjar", ".kt", ".java"],
         ),
+        "common_srcs": attr.label_list(
+            doc = """The list of common files that will be used as common module in multiplatform project. Can only contain Kotlin files.
+                           If `common_srcs` is not null, compiler flags `-Xmulti-platform` will be added to enable multi platform support.""",
+            default = [],
+            allow_files = [".kt"],
+        ),

--- a/kotlin/internal/jvm/compile.bzl
+++ b/kotlin/internal/jvm/compile.bzl
@@ -445,14 +445,15 @@
 def _run_kt_builder_action(
         ctx,
         mnemonic,
         rule_kind,
         toolchains,
         srcs,
         generated_src_jars,
         compile_deps,
         deps_artifacts,
         annotation_processors,
         transitive_runtime_jars,
         plugins,
         outputs,
-        build_kotlin = True):
+        build_kotlin = True,
+        common_srcs = []):
@@ -472,1 +463,4 @@
     args.add_all("--kotlin_passthrough_flags", kotlinc_options_to_flags(kotlinc_options))
+    if len(common_srcs) > 0:
+        args.add("-Xmulti-platform", "-Xexpect-actual-classes")
+    args.add_all(common_srcs, format_each = "-Xcommon-sources=%s")
@@ -659,1 +663,2 @@
         srcs = srcs,
+        common_srcs = ctx.files.common_srcs,
@@ -759,15 +765,16 @@
 def _run_kt_java_builder_actions(
         ctx,
         rule_kind,
         toolchains,
         srcs,
+        common_srcs,
         generated_kapt_src_jars,
         generated_ksp_src_jars,
         compile_deps,
         deps_artifacts,
         annotation_processors,
         ksp_annotation_processors,
         transitive_runtime_jars,
         plugins,
         compile_jar,
         output_jdeps):
@@ -849,15 +855,16 @@
         _run_kt_builder_action(
             ctx = ctx,
             rule_kind = rule_kind,
             toolchains = toolchains,
             srcs = srcs,
+            common_srcs = common_srcs,
             generated_src_jars = generated_kapt_src_jars + generated_ksp_src_jars,
             compile_deps = compile_deps,
             deps_artifacts = deps_artifacts,
             annotation_processors = [],
             transitive_runtime_jars = transitive_runtime_jars,
             plugins = plugins,
             outputs = outputs,
             build_kotlin = True,
             mnemonic = "KotlinCompile",
         )
